File: app/src/main/java/com/example/mytask/MainActivity.java
package com.example.mytask;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.content.Context;
import android.content.Intent;
import android.os.AsyncTask;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.TextView;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity {
RecyclerView recyclerview;
TextView cart_view;
//    Button view_cart;

    int totcount=0;
    CartListAdapter cartListAdapter;
    List<Productfile> productList=new ArrayList<>();
    SessionManager sessionManager;
    private List<Productfile> productfileList;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        cart_view=findViewById(R.id.cart_view);
        recyclerview=findViewById(R.id.recyclerview);

        cart_view.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent=new Intent(MainActivity.this,ViewCart.class);
                startActivity(intent);
            }
        });
        getSupportActionBar().hide();

    }
    private void Productlistmethod() {
        if (productfileList!=null && productfileList.size()>0){
            cartListAdapter=new CartListAdapter(this,productfileList);
            LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this);
            recyclerview.setLayoutManager(linearLayoutManager);
            recyclerview.setAdapter(cartListAdapter);
        }else {
            try {
                JSONObject obj = new JSONObject(loadJSONFromAsset(this));
                JSONArray jsonArray = obj.getJSONArray("users");
                for (int i = 0; i < jsonArray.length(); i++) {
                    JSONObject jsonObject = jsonArray.getJSONObject(i);
                    String pId = jsonObject.getString("PId");
                    String firstName = jsonObject.getString("firstName");
                    String lastName = jsonObject.getString("lastName");
                    String quantity = jsonObject.getString("quantity");
                    Integer price=jsonObject.getInt("price");
                    Productfile productfile = new Productfile();
                    productfile.setCount(0);
                    productfile.setDescription(lastName);
                    productfile.setName(firstName);
                    productfile.setQuanitiy(quantity);
                    productfile.setPid(pId);
                    productfile.setPrice(price);
                    Insertintolocal(productfile);
                    productList.add(productfile);
                  Log.e("productsize",productList.size()+"  ");

                }
                cartListAdapter = new CartListAdapter(this, productList);
                LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this);
                recyclerview.setLayoutManager(linearLayoutManager);
                recyclerview.setAdapter(cartListAdapter);
            } catch (JSONException e) {
                e.printStackTrace();
            }
        }
    }

    private void getProductdetails() {
        class GetTasks extends AsyncTask<Void, Void, List<Productfile>> {

            @Override
            protected List<Productfile> doInBackground(Void... voids) {
               productfileList = DatabaseClient
                        .getInstance(MainActivity.this)
                        .getAppDatabase()
                        .taskDao()
                        .getAllProducts();
                return productfileList;
            }

            @Override
            protected void onPostExecute(List<Productfile> tasks) {
                super.onPostExecute(tasks);
                Productlistmethod();

            }
        }

        GetTasks gt = new GetTasks();
        gt.execute();
    }


    private void Insertintolocal(final Productfile productfile) {
            class SaveTask extends AsyncTask<Void, Void, Void> {

                @Override
                protected Void doInBackground(Void... voids) {

                    DatabaseClient.getInstance(MainActivity.this).getAppDatabase()
                            .taskDao()
                            .insert(productfile);
                    return null;
                }

                @Override
                protected void onPostExecute(Void aVoid) {
                    super.onPostExecute(aVoid);
                }
            }

            SaveTask st = new SaveTask();
            st.execute();

    }

    public static String loadJSONFromAsset(Context context) {
        String json = null;
        try {
            InputStream is = context.getAssets().open("sample_json.json");
            int size = is.available();
            byte[] buffer = new byte[size];
            is.read(buffer);
            is.close();
            json = new String(buffer, "UTF-8");
        } catch (IOException ex) {
            ex.printStackTrace();
            return null;
        }
        return json;
    }

    public void AddcartValue(String count, String action, Productfile productfile) {
        if (action.equalsIgnoreCase("add")){
            totcount= totcount+1;
            cart_view.setText("View Items("+String.valueOf(totcount)+")");

            StoredinLocal(action,productfile,totcount,count);

        }else if (action.equalsIgnoreCase("minus")) {
            if(count.equals("0")){
                totcount=0;
                StoredinLocal(action,productfile,totcount, count);
            }else

                totcount = totcount - 1;
                cart_view.setText("View Items("+String.valueOf(totcount)+")");
                StoredinLocal(action,productfile,totcount, count);

        }

    }

    private void StoredinLocal(String action, final Productfile productfile, final int totcount, final String count) {
        Log.e("StoredinLocal","Working ");
        Log.e("StoredinLocal",count+"  Working ");

        class UpdateTask extends AsyncTask<Void, Void, Void> {

            @Override
            protected Void doInBackground(Void... voids) {
                DatabaseClient.getInstance(getApplicationContext()).getAppDatabase()
                        .taskDao()
                        .updateProduct(Integer.parseInt(count), productfile.getName(), productfile.getPid(),Integer.parseInt(count));
                return null;
            }

            @Override
            protected void onPostExecute(Void aVoid) {
                super.onPostExecute(aVoid);

            }
        }
        UpdateTask st = new UpdateTask();
        st.execute();
    }


    @Override
    protected void onResume() {
        super.onResume();
        getProductdetails();



    }
}

File: app/src/main/java/com/example/mytask/MainActivity.kt
package com.example.mytask

import android.content.Intent
import android.os.AsyncTask
import android.os.Bundle
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.mytask.RoomDB.DatabaseClient
import com.example.mytask.adapters.CartListAdapter
import com.example.mytask.model.Productfile
import java.util.*

class MainActivity : AppCompatActivity() {
    lateinit var recyclerview: RecyclerView
    lateinit var cart_view: TextView
    lateinit var back: ImageView
    var totcount = 0
    lateinit var cartListAdapter: CartListAdapter
    var productList: List<Productfile> = ArrayList()
    private lateinit var productfileList: List<Productfile>
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        cart_view = findViewById(R.id.cart_view)
        back = findViewById(R.id.back)
        recyclerview = findViewById(R.id.recyclerview)
        cart_view.setOnClickListener(View.OnClickListener {
            val intent = Intent(this@MainActivity, ViewCart::class.java)
            startActivity(intent)
        })
        supportActionBar!!.hide()
        back.setOnClickListener(View.OnClickListener { finish() })

    }

    private fun productlistmethod() {
        if (productfileList != null && productfileList!!.size > 0) {
            cartListAdapter = CartListAdapter(this, productfileList)
            val linearLayoutManager = LinearLayoutManager(this)
            recyclerview!!.layoutManager = linearLayoutManager
            recyclerview!!.adapter = cartListAdapter
        }
    }

    private val productdetails: Unit
        private get() {
            class GetTasks : AsyncTask<Void?, Void?, List<Productfile>>() {
                override fun doInBackground(vararg params: Void?): List<Productfile>  {
                    productfileList = DatabaseClient
                            .getInstance(this@MainActivity)
                            ?.appDatabase
                            ?.taskDao()
                            ?.getAllProducts()!!
                    return productfileList
                }

                override fun onPostExecute(tasks: List<Productfile>?) {
                    super.onPostExecute(tasks)
                    productlistmethod()
                }

            }

            val gt = GetTasks()
            gt.execute()
        }

    fun AddcartValue(count: String, action: String, productfile: Productfile) {
        if (action.equals("add", ignoreCase = true)) {
            totcount = totcount + 1
            cart_view!!.text = "View Items($totcount)"
            StoredinLocal(productfile, count)
        } else if (action.equals("minus", ignoreCase = true)) {
            if (count == "0") {
                totcount = 0
                StoredinLocal(productfile, count)
            } else totcount = totcount - 1
            cart_view!!.text = "View Items($totcount)"
            StoredinLocal(productfile, count)
        }
    }

    private fun StoredinLocal(productfile: Productfile, count: String) {
        class UpdateTask : AsyncTask<Void?, Void?, Void?>() {
            override fun doInBackground(vararg params: Void?): Void? {
                DatabaseClient.getInstance(applicationContext)?.appDatabase
                        ?.taskDao()
                        ?.updateProduct(count.toInt(), productfile.name, productfile.regid, count.toInt())
                return null
            }

            override fun onPostExecute(aVoid: Void?) {
                super.onPostExecute(aVoid)
            }

        }

        val st = UpdateTask()
        st.execute()
    }

    override fun onResume() {
        super.onResume()
        productdetails
    }
}